backupccl: altering backup schedule no longer resumes schedule

kev-cao opened the pull request:
Previously, altering a backup schedule's recurrence or collection URI would resume the schedule even if it was paused. As we already have `RESUME SCHEDULE`, this is not necessary and can lead to unexpected surprises for the user. The new behavior is upon altering a paused schedule, the schedules remain paused regardless of if the collection URI or the recurrence is changed.

However, in the event that the collection URI is altered for a pair of active schedules, the incremental will still be paused until a full backup is fully completed.

Fixes: #128995

Epic: None

Release note (backward-incompatible change): Altering a paused backup schedule's recurrence or location no longer resumes the schedule.

kev-cao committed:
backupccl: altering backup schedule no longer resumes schedule

Previously, altering a backup schedule's recurrence or collection URI
would resume the schedule even if it was paused. As we already have
`RESUME SCHEDULE`, this is not necessary and can lead to unexpected
surprises for the user. The new behavior is upon altering a paused
schedule, the schedules remain paused regardless of if the collection
URI or the recurrence is changed.

However, in the event that the collection URI is altered for a pair of
active schedules, the incremental will still be paused until a full
backup is fully completed.

Fixes: #128995

Epic: None

Release note (backward-incompatible change): Altering a paused backup
schedule's recurrence or location no longer resumes the schedule.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134829)
<!-- Reviewable:end -->

msbutler commented on pkg/backup/alter_backup_schedule.go:
```diff
@@ -399,13 +399,18 @@ func processRecurrence(
+		if fullJob.IsPaused() {
+			fullJob.SetScheduleExpr(recurrence)
+		} else {
+			return fullJob.SetSchedule(recurrence)
```
nit: rename `SetSchedule` to `SetScheduleAndNextRun`

msbutler commented on pkg/backup/alter_backup_schedule.go:
```diff
@@ -573,14 +580,18 @@ func processInto(p sql.PlanHookState, spec *alterBackupScheduleSpec, s scheduleD
-	// This mirrors the behavior of CREATE SCHEDULE FOR BACKUP.
-	env := sql.JobSchedulerEnv(p.ExecCfg().JobsKnobs())
-	s.fullJob.SetNextRun(env.Now())
+	// If the schedule was not already paused, kick off a full backup immediately
```
s/if the schedule/if the inc schedule/r

msbutler commented on pkg/backup/alter_backup_schedule_test.go:
```diff
@@ -194,3 +194,88 @@ func TestAlterBackupScheduleSetsIncrementalClusterID(t *testing.T) {
+	defer cleanup()
+
+	t.Run("standalone paused full backup is not resumed", func(t *testing.T) {
+		rows := th.sqlDB.QueryStr(t,
```
nit: this is a cleaner way to grab the schedule id:
```
var scheduleID int
QueryRow(stmt).Scan(&scheduleID)
```

kev-cao replied:
If I'm not mistaken, `QueryRow().Scan` requires you to provide a location for every column in the row or else it errors. Although tbh I only did it this way because I copied it from above.

msbutler commented on pkg/backup/alter_backup_schedule_test.go:
```diff
@@ -194,3 +194,88 @@ func TestAlterBackupScheduleSetsIncrementalClusterID(t *testing.T) {
+		th.sqlDB.Exec(
+			t,
+			fmt.Sprintf(
+				`ALTER BACKUP SCHEDULE %d SET RECURRING '@hourly',
```
nit: bind the cmd to a local var so the Exec cmd can be one line.

msbutler commented on pkg/backup/alter_backup_schedule_test.go:
```diff
@@ -194,3 +194,88 @@ func TestAlterBackupScheduleSetsIncrementalClusterID(t *testing.T) {
+				scheduleID,
+			))
+
+		require.Equal(t, "PAUSED", getScheduleStatus(t, th, scheduleID))
```
check that the recurrence changed?

msbutler commented on pkg/backup/alter_backup_schedule_test.go:
```diff
@@ -194,3 +194,88 @@ func TestAlterBackupScheduleSetsIncrementalClusterID(t *testing.T) {
+		rows := th.sqlDB.QueryStr(t,
+			`CREATE SCHEDULE FOR BACKUP INTO 'nodelocal://1/backup-2/alter-schedule' RECURRING '@hourly' FULL BACKUP '@daily';`)
+		require.Len(t, rows, 2)
+		scheduleID1, err := strconv.Atoi(rows[0][0])
```
heh, this may be cleaner than using sqlDB.Query.

msbutler commented on pkg/backup/alter_backup_schedule_test.go:
```diff
@@ -194,3 +194,88 @@ func TestAlterBackupScheduleSetsIncrementalClusterID(t *testing.T) {
+		th.sqlDB.Exec(
+			t,
+			fmt.Sprintf(
+				`ALTER BACKUP SCHEDULE %d SET RECURRING '*/30 * * * *',
```
bind cmd to local var

msbutler commented on pkg/backup/alter_backup_schedule_test.go:
```diff
@@ -194,3 +194,88 @@ func TestAlterBackupScheduleSetsIncrementalClusterID(t *testing.T) {
+		)
+
+		require.Equal(t, "PAUSED", getScheduleStatus(t, th, scheduleID1))
+		require.Equal(t, "PAUSED", getScheduleStatus(t, th, scheduleID2))
```
check recurrence?

msbutler commented on pkg/backup/alter_backup_schedule_test.go:
```diff
@@ -194,3 +194,88 @@ func TestAlterBackupScheduleSetsIncrementalClusterID(t *testing.T) {
+func getScheduleStatus(t *testing.T, th *alterSchedulesTestHelper, id int) string {
+	t.Helper()
+	var status string
+	th.sqlDB.QueryRow(t, `SELECT schedule_status FROM [SHOW SCHEDULES] WHERE id=$1`, id).Scan(&status)
```
ah you could easily grab the new recurrence her via `Scan(&status,&recurrence)`

kev-cao replied:
Yup, updated!

msbutler commented on pkg/backup/alter_backup_schedule_test.go:
```diff
@@ -194,3 +194,88 @@ func TestAlterBackupScheduleSetsIncrementalClusterID(t *testing.T) {
+		fmt.Sprintf(`ALTER BACKUP SCHEDULE %d SET INTO 'nodelocal://1/backup/alter-schedule-2';`, fullSchedID),
+	)
+
+	require.Equal(t, "PAUSED", getScheduleStatus(t, th, incSchedID))
```
this isn't a change in behavior right?  If, however, the schedules were paused before altering the uri, do they stay paused?

kev-cao replied:
> the schedules were paused before altering the uri, do they stay paused?

And yes, they do now with this change

msbutler replied:
i'd split out this additional test into a separate commit.

kev-cao replied:
Sounds good, updated!

msbutler commented on pkg/backup/alter_backup_schedule.go:
```diff
@@ -433,9 +438,7 @@ func processFullBackupRecurrence(
-		if err := s.fullJob.SetSchedule(s.incJob.ScheduleExpr()); err != nil {
-			return scheduleDetails{}, err
-		}
+		s.fullJob.SetScheduleExpr(s.incJob.ScheduleExpr())
```
i'm having a really hard time understanding what `processFullBackupRecurrence` does. I would have expected it to change the `FULL BACKUP crontab`, but it has turned into a dirty kitchen sink. In a follow up pr, could you clean this logic up so mortals can understand what's going on? This clean up work is less important than backup compaction stuff.

kev-cao replied:
Yeah for sure, it's definitely got a lot going on

msbutler reviewed (commented):
nice!

kev-cao committed:
backupccl: add test for alter backup schedule into new URI

We previously were missing test coverage for ensuring that upon altering
the collection URI for *active* schedules, the incremental schedule is
paused, waiting until the full backup completes.

Epic: none

Release note: none

kev-cao commented:
TFTR!

bors r=msbutler
